sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,203 @@
+
+        """
+        return pika.BasicProperties(content_type, content_encoding, headers,
+                                    delivery_mode, priority, correlation_id,
+                                    reply_to, expiration, message_id,
+                                    timestamp, message_type, user_id, app_id)
+
+    def _new_rabbitmq_connection(self):
+        """Return a connection to RabbitMQ via the pika.Connection object.
+        When RabbitMQ is connected, on_rabbitmq_open will be called.
+
+        :rtype: pika.adapters.tornado_connection.TornadoConnection
+
+        """
+        return tornado_connection.TornadoConnection(self._rabbitmq_parameters,
+                                                    self.on_conn_open)
+
+    def _publish_message(self, exchange, routing_key, message, properties):
+        """Publish the message to RabbitMQ
+
+        :param str exchange: The exchange to publish to
+        :param str routing_key: The routing key to publish with
+        :param str message: The message body
+        :param pika.BasicProperties: The message properties
+
+        """
+        global CHANNEL
+        LOGGER.debug('Publishing message: %s to exchange: %s with '
+                     'routing_key: %s and properties: %s',
+                     message, exchange, routing_key, properties)
+        if self._rabbitmq_is_closed or not CHANNEL:
+            LOGGER.error('Unable to connect to Rabbit')
+            return
+        CHANNEL.basic_publish(exchange, routing_key, message, properties)
+
+    @property
+    def _rabbitmq_is_closed(self):
+        """Returns True if the pika connection to RabbitMQ is closed.
+
+        :rtype: bool
+
+        """
+        return not CONNECTION
+
+    @property
+    def _rabbitmq_parameters(self):
+        """Return a pika URLParameters object using the :envvar:`AMQP`
+        environment variable which identifies the Rabbit MQ server as a URL
+        ready for :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        LOGGER.info('URLParameters: %s', pika.URLParameters(os.environ['AMQP']))
+        return pika.URLParameters(os.environ['AMQP'])
+
+    def on_conn_close(self, reply_code, reply_text):
+        """Called when RabbitMQ has been connected to.
+
+        :param int reply_code: The code for the disconnect
+        :param str reply_text: The disconnect reason
+
+        """
+        global CONNECTION, CHANNEL
+        LOGGER.warning('RabbitMQ has disconnected (%s): %s',
+                       reply_code, reply_text)
+        CONNECTION = None
+        CHANNEL = None
+        self._connect_to_rabbitmq()
+
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        global CONNECTION, CHANNEL
+        LOGGER.info('RabbitMQ is attempting to connect')
```

Comment:
we've already connected, we are fetching a channel of communication at this point.